[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73528

Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

dzdang committed:
[Quant][fx] Added explicit entrieentries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"




[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"




[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"




[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -253,19 +253,38 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
 
-def get_default_qconfig_dict(backend='fbgemm', version=0):
-    qconfig = get_default_qconfig(backend)
+def get_default_qconfig_dict_helper(qconfig, qconfig_transpose):
```
nit: let's add an underscore to show this is not public facing

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -253,19 +253,38 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
+                        (torch.nn.functional.linear, qconfig)]}
+
+def get_default_qconfig_dict(backend='fbgemm', version=0):
+    qconfig = get_default_qconfig(backend)
+    qconfig_transpose = qconfig
```
maybe we can add a version argument for `get_default_qconfig` as well

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -253,19 +253,38 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
+def get_default_qconfig_dict(backend='fbgemm', version=0):
+    qconfig = get_default_qconfig(backend)
+    qconfig_transpose = qconfig
+    if backend == "fbgemm":
```
can we add a comment explaining why we need a different qconfig for convtranspose?

jerryzh168 commented:
Please add a Summary and Test for this PR

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -235,8 +239,8 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
-    # Use the fused observer + fake_quant modules for doing QAT.
-    if version == 1:
+    # Use the fused observe + fake_quant modules for doing QAT.
+    elif version == 1:
```
should we add a raise here as well?

jerryzh168 replied:
sorry, missed the else below

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -251,21 +255,50 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
+def get_default_qconfig_dict(backend='fbgemm', version=0):
+    qconfig = get_default_qconfig(backend, version)
+    qconfig_transpose = qconfig
+    # default_per_channel_weight_observer is not currently compatible with fbgemm backend
```
we can add a link to this issue here as well I think: https://github.com/pytorch/pytorch/issues/47535

dzdang replied:
added

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -251,21 +255,50 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
+    # so we have to modify the weight observer to MovingAverageMinMaxObserver or another
+    # per tensor supported observer
+    if backend == "fbgemm":
+        qconfig_transpose = QConfig(activation=qconfig.activation, weight=MovingAverageMinMaxObserver)
```
we need a fake_quant for weight as well for qat, maybe change this to `default_weight_fake_quant`?

jerryzh168 commented on torch/ao/quantization/qconfig.py:
```diff
@@ -251,21 +255,50 @@ def get_default_qat_qconfig(backend='fbgemm', version=1):
+    # so we have to modify the weight observer to MovingAverageMinMaxObserver or another
+    # per tensor supported observer
+    if backend == "fbgemm":
+        qconfig_transpose = QConfig(activation=qconfig.activation, weight=MovingAverageMinMaxObserver)
```
I think the weight observer here can be: `default_weight_observer`

jerryzh168 reviewed (approved):
looks good, thanks! had a few inline comments

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).

dzdang committed:
Update on "[Quant][fx] Added explicit entries for for functional and module conv&linear support into get_default_qconfig_dict&get_default_qat_qconfig_dict"


Differential Revision: [D34535572](https://our.internmc.facebook.com/intern/diff/D34535572)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34535572).
